src/client/components/PasscodeInput.tsx

```diff
@@ -30,6 +65,9 @@ export const PasscodeInput = ({
 				error={fieldErrors?.find((error) => error.field === 'code')?.message}
 				defaultValue={passcode}
 				cssOverrides={passcodeInputStyles}
+				value={input.value}
+				onChange={onChange}
+				onPaste={onPaste}
```

Comment:
A paste event should trigger `onChange`, why is it needed separately?

Reply:
I suppose that's true, but having a paste event keeps it more flexible if we were to for example add an auto submit functionality.

Reply:
My concern is that we're needlessly re-rendering twice on paste events - it's a tiny issue really.

Reply:
Also, I don't think we'd want an auto-submit `onPaste`. I think we'd want an auto-submit when `input.value.length === 6`.

Reply:
I've added the `onPaste` back in, as there was strange behaviour when pasting when only using `onChange`.

Essentially lets say your clipboard has an extra space or another at the start of the string by accident, e.g. ` 123456`. Then when pasting in when only using `onChange` the input value is left at `12345` for some reason, as it seems to only look up to the 6th character before sanitising. When using `onPaste` this doesn't happen and the correct input value of `123456` is appled.

Reply:
Oh, that's super weird. I think that's a good enough reason to keep it in!

Reply:
I wonder if it's something to do with the html validation interfering with the javascript/react validation?

either way, it's fine as it is, gateway doesn't use do much rerendering at all